File: nobodywho/testing-apps/kotlin/src/androidTest/java/ai/nobodywho/testapp/DeviceInferenceTest.kt
package ai.nobodywho.testapp

import ai.nobodywho.Chat
import ai.nobodywho.Message
import ai.nobodywho.Model
import ai.nobodywho.Tool
import ai.nobodywho.text
import androidx.test.ext.junit.runners.AndroidJUnit4
import androidx.test.platform.app.InstrumentationRegistry
import kotlinx.coroutines.flow.toList
import kotlinx.coroutines.runBlocking
import org.junit.Assert.assertEquals
import org.junit.Assert.assertFalse
import org.junit.Assert.assertNotNull
import org.junit.Test
import org.junit.runner.RunWith

// Top-level function so Tool's KFunction reflection can bind to it. Tool only
// supports top-level / class / companion functions, not local ones.
fun ping(): String = "pong"

/**
 * On-device smoke test that runs on real hardware via Firebase Test Lab.
 *
 * Mirrors the chat / streaming / tool-calling assertions from the host-JVM
 * `IntegrationTest`, but exercises the arm64 Android `.so` on a physical phone.
 *
 * CI points `-e modelUrl` at an OBB it preloaded, so the device never downloads
 * the model. A local run takes the default and fetches it from Hugging Face.
 */
@RunWith(AndroidJUnit4::class)
class DeviceInferenceTest {

    // Overridable via `-e modelUrl <url-or-path>` so the workflow can point at a
    // preloaded model without recompiling; defaults to what the JVM tests use.
    private fun modelUrl(): String =
        InstrumentationRegistry.getArguments().getString("modelUrl")
            ?: "hf://NobodyWho/Qwen_Qwen3-0.6B-GGUF/Qwen_Qwen3-0.6B-Q4_K_M.gguf"

    @Test
    fun chatCompletesStreamsAndCallsTools() = runBlocking {
        // Ask for the GPU like a real app would. Android has no GPU backend
        // yet, so this falls back to CPU today and starts exercising the GPU
        // path automatically once one lands.
        val model: Model = Model.load(modelUrl(), useGpu = true)

        // Completion
        val chat = Chat(
            model = model,
            systemPrompt = "Reply with one word only.",
            templateVariables = mapOf("enable_thinking" to false),
        )
        val response: String = chat.ask("Say hello").completed()
        assertFalse("Completion should be non-empty", response.isEmpty())

        // Streaming
        chat.resetContext(systemPrompt = "Reply briefly.")
        val tokens: List<String> = chat.ask("Say hi").asFlow().toList()
        assertFalse("Streaming should yield at least one token", tokens.isEmpty())

        // Tool calling
        val pingTool: Tool = Tool(
            name = "ping",
            description = "Ping the server",
            function = ::ping,
        )
        chat.resetContext(
            systemPrompt = "Use the ping tool now.",
            tools = listOf(pingTool),
        )
        chat.ask("Ping the server").completed()
        val toolResponse: Message.Tool? =
            chat.getChatHistory().filterIsInstance<Message.Tool>().firstOrNull()
        assertNotNull("Expected a tool response in chat history", toolResponse)
        val toolText: String = toolResponse!!.content.text
        assertEquals("pong", toolText)
    }
}

File: nobodywho/testing-apps/flutter/integration_test/device_inference_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:integration_test/integration_test.dart';
import 'package:logging/logging.dart';
import 'package:nobodywho/nobodywho.dart' as nobodywho;

/// The model is downloaded on-device into the app's own cache on first run —
/// no permissions and no shared storage involved.
const modelUrl =
    'hf://NobodyWho/Qwen_Qwen3-0.6B-GGUF/Qwen_Qwen3-0.6B-Q4_K_M.gguf';

String ping() => 'pong';

/// On-device smoke test, run on real hardware via Firebase Test Lab.
///
/// Mirrors the Kotlin binding's `DeviceInferenceTest`: completion, streaming
/// and tool calling, exercising the arm64 native library on a physical phone.
void main() {
  IntegrationTestWidgetsFlutterBinding.ensureInitialized();

  testWidgets(
    'chat completes, streams and calls tools',
    (tester) async {
      // FRB forwards Rust/llama.cpp logs to Dart logging, but the library
      // intentionally leaves output to its host. Print them into FTL logcat.
      final previousLogLevel = Logger.root.level;
      Logger.root.level = Level.ALL;
      final logSubscription = Logger.root.onRecord.listen((record) {
        // ignore: avoid_print
        print(
          '[${record.level.name}] ${record.loggerName}: ${record.message}'
          '${record.error == null ? '' : '\n${record.error}'}'
          '${record.stackTrace == null ? '' : '\n${record.stackTrace}'}',
        );
      });
      addTearDown(() async {
        await logSubscription.cancel();
        Logger.root.level = previousLogLevel;
      });
      await nobodywho.NobodyWho.init();

      // Exercise automatic OpenCL/Vulkan selection with CPU fallback.
      final chat = await nobodywho.Chat.fromPath(
        modelPath: modelUrl,
        systemPrompt: 'Reply with one word only.',
        templateVariables: const {'enable_thinking': false},
        useGpu: true,
      );

      // Completion
      final response = await chat.ask('Say hello').completed();
      expect(response, isNotEmpty, reason: 'completion should be non-empty');

      // Streaming
      await chat.resetContext(systemPrompt: 'Reply briefly.', tools: []);
      var tokenCount = 0;
      await for (final _ in chat.ask('Say hi')) {
        tokenCount++;
      }
      expect(tokenCount, greaterThan(0),
          reason: 'streaming should yield at least one token');

      // Tool calling. The Tool is constructed here rather than up front because
      // handing one to the Rust side consumes it — reusing a Tool across calls
      // currently throws DroppableDisposedException (NOB-168).
      final pingTool = nobodywho.Tool(
        function: ping,
        name: 'ping',
        description: 'Ping the server',
      );
      await chat.resetContext(
        systemPrompt: 'Use the ping tool now.',
        tools: [pingTool],
      );
      await chat.ask('Ping the server').completed();

      final toolMessages =
          (await chat.getChatHistory()).whereType<nobodywho.Message_Tool>();
      expect(toolMessages, isNotEmpty,
          reason: 'expected a tool response in chat history');
      expect(toolMessages.first.content.text, 'pong');
    },
    // Generous: the first run downloads the model before any inference starts.
    timeout: const Timeout(Duration(minutes: 20)),
  );
}
